gapis/trace/android/profile/profile.go

```diff
@@ -134,77 +143,117 @@ func setGpuCounterMetrics(ctx context.Context, groupToSlices map[int32][]*servic
-			ts = append(ts, counter.Timestamps[i])
-			vs = append(vs, counter.Values[i])
+// Scan global slices and count concurrent slices for each counter sample.
+func scanConcurrency(globalSlices []*service.ProfilingData_GpuSlices_Slice, counter *service.ProfilingData_Counter) []int {
+	slicesCount := make([]int, len(counter.Timestamps))
+	for _, slice := range globalSlices {
+		sStart, sEnd := slice.Ts, slice.Ts+slice.Dur
+		for i := 1; i < len(counter.Timestamps); i++ {
+			cStart, cEnd := counter.Timestamps[i-1], counter.Timestamps[i]
+			if cEnd < sStart { // Sample earlier than GPU slice's span.
+				continue
+			} else if cStart > sEnd { // Sample later than GPU slice's span.
+				break
+			} else { // Sample overlaps with GPU slice's span.
+				slicesCount[i]++
+			}
 		}
 	}
-	if len(ts) == 0 {
-		return float64(-1)
-	}
+	return slicesCount
+}
 
-	// Aggregate counter samples.
-	// Contribution time is the overlapped time between a counter sample's implicit range and a gpu slice.
-	ctSum := uint64(0)        // Accumulation of contribution time.
-	weightedSum := float64(0) // Accumulation of (counter value * counter's contribution time).
+// Map counter samples to GPU slice. When collecting samples, three sets will
+// be maintained based on attribution strategy: the minimum set,
+// the best guess set, and the maximum set.
+// The returned results map {sample index} to {sample weight}.
+func mapCounterSamples(slices []*service.ProfilingData_GpuSlices_Slice, counter *service.ProfilingData_Counter, concurrentSlicesCount []int) (map[int]float64, map[int]float64, map[int]float64) {
+	estimateSet, minSet, maxSet := map[int]float64{}, map[int]float64{}, map[int]float64{}
 	for _, slice := range slices {
 		sStart, sEnd := slice.Ts, slice.Ts+slice.Dur
-		if ts[0] > sStart {
-			ct := u64.Min(ts[0], sEnd) - sStart
-			ctSum += ct
-			weightedSum += float64(ct) * vs[0]
-		}
-		for i := 1; i < len(ts); i++ {
-			cStart, cEnd := ts[i-1], ts[i]
+		for i := 1; i < len(counter.Timestamps); i++ {
+			cStart, cEnd := counter.Timestamps[i-1], counter.Timestamps[i]
+			concurrencyWeight := 1.0
+			if concurrentSlicesCount[i] > 1 {
+				concurrencyWeight = 1 / float64(concurrentSlicesCount[i])
+			}
 			if cEnd < sStart { // Sample earlier than GPU slice's span.
 				continue
-			} else if cEnd < sEnd { // Sample inside GPU slice's span, or sample's latter part overlaps with slice.
-				ct := cEnd - u64.Max(cStart, sStart)
-				ctSum += ct
-				weightedSum += float64(ct) * vs[i]
-			} else if cStart < sEnd { // Sample wraps GPU slice's span, or sample's earlier part overlaps with slice.
-				ct := sEnd - u64.Max(sStart, cStart)
-				ctSum += ct
-				weightedSum += float64(ct) * vs[i]
+			} else if cStart > sEnd { // Sample later than GPU slice's span.
 				break
+			} else if cStart > sStart && cEnd < sEnd { // Sample is contained inside GPU slice's span.
+				estimateSet[i] = 1 * concurrencyWeight
+				minSet[i] = 1 * concurrencyWeight
```

Comment:
I think the minimum should only include this value, if there are no concurrent slices. If there are concurrent slices, there is a possibility that the value is entirely attributable to one of the other concurrent slices.

Reply:
Good catch! Done in a new commit.